### PR TITLE
Introduce Field & DataSet to eliminate sub-trainers & sub-testers

### PR DESCRIPTION
**Pull request by FengZiYjun**
### Field, Instance & DataSet
`dataset.py` defines `DataSet`, which is a list of `Instance`s.
`instance.py` defines `Instance`, which is a single example and contains multiple `Field`s.

`field.py` defines `Field`, which is the elementary data type or representation.
`TextField` defines a list of strings. `LabelField` defines single interger or string.
You can add extra fields to support more complex data.

Each field
-  has a field name
-  has a `is_target` boolean argument to specify whether it is Y or not (X) in training.
-  has a `to_tensor` method to define how this field data is transformed into tensors

`dataset.py` defines a function to make DataSet from a list.
```python
def create_dataset_from_lists(str_lists: list, word_vocab: dict, has_target: bool = False, label_vocab: dict = None) --> DataSet: 
```
Example: https://github.com/fastnlp/fastNLP/blob/ad044ef4c76c2c52da6e732a67ff2001e7a677d5/test/core/test_tester.py#L15
https://github.com/fastnlp/fastNLP/blob/ad044ef4c76c2c52da6e732a67ff2001e7a677d5/test/core/test_trainer.py#L14

### Batch
`batch.py` defines `Batch`, an iterable wrapper of `DataSet`.
Sampling and padding is applied insides.
Iteration over a `Batch` object returns two dict, `batch_x` and `batch_y`. 
The key of the dict is the field name. The value is the corresponding batch tensor.

 ```Python
data_iterator = Batch(data_set, batch_size=8, sampler=RandomSampler(), use_cuda=False)
for batch_x, batch_y in data_iterator:
    batch_x["word_seq"]  # torch.LongTensor
    batch_y["label"]  # torch.LongTensor
```

#### how to get the pre-paded (origin) length of the sequence
`Batch` will keep a record of the origin length of a field before padding. 
It returns the origin lengths with a string key created by appending "_origin_len" before the field name.
For example,  `batch_x["word_seq_origin_len"]  # torch.LongTensor`.

*Why origin length is tensor rather than a list of int ?*
*Because the sequence labeling model's forward() has added an extra arguemnt `seq_len` to represent the origin lengths (The creation of sequence masks is moved into the model, which needs `seq_len`.). And tensorboardX requires arguemnts passed to forward() to be nothing but tensor.*

### One Trainer, Many Tasks
In previous design, different trainers are responsible for different tasks.
After introducing Fields & DataSet, different tasks are represented by different DataSet structures, which is the way Fields organize.
Therefore, all methods in `SeqLabelTrainer` and  `ClassificationTrainer` are removed. They are just an empty sub-class to deprecate, and will throw an warning information when used.
So are those in Testers and Predictor.

#### self._task for trainers, testers, and predictors
However, trainers still need task information to know **which fields are network inputs among batch_x**.
This is because 
- we don't know which task is going to do when preprocessing and making DataSet.
- not all fields in batch_x are needed as network input. Some may be unused, such as `seq_len` in text classification.
- in tester, different tasks require different evaluation methods and metrics.
- in predictor, different tasks require different pre-process and post-process.

Trainer & Tester have a **required** arguement (raise error if not provided, NO default value) `self._task` to specify which task is going to perform.

```python
if self._task == "seq_label": 
    y = network(x["word_seq"], x["word_seq_origin_len"]) 
elif self._task == "text_classify": 
    y = network(x["word_seq"]) 
```

### How to add your task ?
1. design a pytorch model, with forward method.
2. choose fields or create new fields to describe your data set.
3. modify preprocessor's `build_dict` method: to build dictionary over your data, and use the dictionary to transform multi-level list of strings into multi-level list of indices.
https://github.com/fastnlp/fastNLP/blob/ef3c753e0db37c710f4068403c6efde4fcb9c3c4/fastNLP/core/preprocess.py#L182
4. modify preprocessor's `convert_to_dataset` method: to transform multi-level list of indices into a DataSet object.
https://github.com/fastnlp/fastNLP/blob/ef3c753e0db37c710f4068403c6efde4fcb9c3c4/fastNLP/core/preprocess.py#L244
5. specify which fields you want to use as network inputs in Trainer, Tester, and Predictor. Where `self._task` appears, where there are modification. 
6. run and debug.

### Future Works
1. optimize Preprocessor: make it a callable object, customized processing function as argument
2. more unit tests on core/
3. eliminate `self._task` ?
4. merge kezhen's code about build_dict

Any questions are welcome!

### CODE REVIEW

**Comment by xpqiu** [comment]
I can first accept it and then check.

**Comment by FengZiYjun** [comment]
continuous integration fails due to the use of command line argument parser in `test_seq_label.py`. 
just delete the parser

File: fastNLP/core/field.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return torch.LongTensor(self._index + pads)
+
+
+class LabelField(Field):
+    def __init__(self, label, is_target=True):
+        super(LabelField, self).__init__(is_target)
+        self.label = label
+        self._index = None
+
+    def get_length(self):
```
Comment (xpqiu) [inline fastNLP/core/field.py]: is the "sequence of labels" also use this LabelField?

Comment (xpqiu) [reply fastNLP/core/field.py]: i see

Comment (FengZiYjun) [reply fastNLP/core/field.py]: No :D
It uses TextField. LabelField is actually single label.
